### PR TITLE
perf(funnels): Deep filtering of person distinct ID join

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

There is a problem in funnels: We currently always read _all_ the person distinct IDs of a project before we join those IDs with events. This baloons the memory requirements of the `events<>person_distinct_ids2` join, acutely affecting customers with larger numbers of distinct IDs – I found the problem based on [this particular report](https://posthog.slack.com/archives/C04R96TUZDE/p1706102503629449).

## Changes

Reading the events or distinct IDs from disk is really fast in itself, what's expensive is making a join with lots of data on the right side. So there is a pretty straightforward remedy: in most cases we can cut down the number of distinct IDs used in the join by only considering _relevant_ distinct IDs – those which actually appear in the set of events. That's what this PR does.

## How did you test this code?

Basically, all tests should continue to pass. Look at SQL snapshots for change to the query's shape.

### CODE REVIEW

File: posthog/queries/funnels/funnel_event_query.py [hunk 1]
```diff
@@ -102,15 +102,16 @@ def get_query(
...
         # but due to ordering requirements in functions building this query we do
         # things like this for now but should do a larger refactor to get rid of it
         query = f"""
-            SELECT {', '.join(_fields)}
-            {{extra_select_fields}}
+            SELECT {all_fields}
             FROM events {self.EVENT_TABLE_ALIAS}
             {sample_clause}
-            {self._get_person_ids_query()}
+            {self._get_person_ids_query(relevant_events_conditions=entity_query + date_query)}
```
Comment (Twixes) [inline posthog/queries/funnels/funnel_event_query.py:114]: Using `prop_query` would be yet another major reduction in memory usage in queries with narrowly defined filters, but I tried it and this is currently pretty painful because suddenly this deep filtering query needs not just conditions, but also fields. I found this to be tricky to get right. Likely will be easier to reason about that when this optimization is part of a HogQL visitor, rather than this query concat monster.

**Review by mariusandra — APPROVED** [review]
Makes sense, though we will need to re-implement for HogQL down the line.. 👍
